ipn/ipnlocal/local.go

```diff
@@ -6203,3 +6204,157 @@ func mayDeref[T any](p *T) (v T) {
 	}
 	return *p
 }
+
+var errNoSuggestion = errors.New("no suggestion available")
+var errUnableToPick = errors.New("unable to pick candidate")
+
+// SuggestExitNode returns a tailcfg.StableNodeID, name, and tailcfg.Location of a suggested exit node given the local backend's netmap and last report.
+// Non-mullvad nodes are prioritized before suggesting Mullvad nodes.
+// We look at non-mullvad nodes region latency values, and if there are none then we choose the geographic closest Mullvad node to the self node's preferred DERP region.
+func (b *LocalBackend) SuggestExitNode() (tailcfg.StableNodeID, string, tailcfg.Location, error) {
+	b.mu.Lock()
+	lastReport := b.MagicConn().GetLastNetcheckReport(b.ctx)
+	netMap := b.netMap
+	b.mu.Unlock()
+	return suggestExitNode(lastReport, netMap)
+}
+
+func suggestExitNode(lastReport *netcheck.Report, netMap *netmap.NetworkMap) (tailcfg.StableNodeID, string, tailcfg.Location, error) {
+	var preferredExitNodeID tailcfg.StableNodeID
+	var preferredExitNodeName string
+	var preferredExitNodeLocation tailcfg.Location
+	if lastReport.PreferredDERP == 0 {
+		return preferredExitNodeID, "", tailcfg.Location{}, errUnableToPick
+	}
+	var candidates []tailcfg.NodeView
+	peers := netMap.Peers
+	for _, peer := range peers {
+		if peer.CapMap().Has(tailcfg.NodeAttrSuggestExitNode) {
+			if online := peer.Online(); online != nil && !*online {
+				continue
+			}
+			if tsaddr.ContainsExitRoutes(peer.AllowedIPs()) {
+				candidates = append(candidates, peer)
+			}
+		}
+	}
+	if len(candidates) == 1 {
+		if candidates[0].Valid() {
+			hi := candidates[0].Hostinfo()
+			if hi.Valid() && hi.Location().View().Valid() {
+				preferredExitNodeLocation = *hi.Location()
+			}
+			return candidates[0].StableID(), candidates[0].Name(), preferredExitNodeLocation, nil
+		} else {
+			return preferredExitNodeID, "", tailcfg.Location{}, errUnableToPick
+		}
+	}
+	if len(candidates) == 0 {
+		return preferredExitNodeID, "", tailcfg.Location{}, errNoSuggestion
+	}
+
+	candidateLatencyMap := make(map[int][]tailcfg.NodeView, len(netMap.DERPMap.Regions))
+	candidateDistanceMap := make(map[float64][]tailcfg.NodeView, len(candidates))
```

Comment:
A map doesn't make sense for this, more likely a slice of distance + nodeview. From there, I can see two approaches:
- Sort and then `pickWeighted` from the first N elements, where N is selected based on a cutoff of the minimum value + some small number to include things that are close enough to be effectively equal. This is O(n log n).
- In a first pass, find the minimum distance. In a second pass, discard all elements that are further away than the minimum distance + some small number. Run `pickWeighted` on the filtered result. This is O(n).